src/main/connection_listener/mod.rs

```diff
@@ -146,11 +162,19 @@ impl<UID: Uid> ConnectionListener<UID> {
         loop {
             match self.listener.accept() {
                 Ok((socket, _)) => {
+                    let mut socket = TcpSock::wrap(socket);
+                    if let Err(e) = socket.set_decrypt_ctx(DecryptContext::anonymous_decrypt(
```

Comment:
This might be tricky. You cannot blanket set this as it depends on what the message is. If someone is directly connecting to us they would use authenticated encryption as crust allows _active_ connections only between known peers. If it's a bootstrap accept it will be anonymous decryption as we don't know the enc_pk of bootstrapper until the handshake.

Maybe there should be a mode in sock-coll to say try-all and report which one was successful. Then in exchg_msg here, enforce that if anon was successful the msg must be Bootstrap-req or IpEcho-req. For a connect-attempt it must be authenticated encryption.

Reply:
Actually, the direct connections will always start with anonymous encryption and then will upgrade to authenticated one. That's because the listener is not aware of who will be connecting: anybody can connect to the same listener. Given that, we should be good in this case with no extra changes/hacks in socket-collection :)

Reply:
In order to initiate an _active_ connection attempt, you are required to know the peer completely. So it will give you authenticated encrypted msg. Only bootstrapping nodes will give you anon as you as a proxy are not expected to know the pks of all the future bootstrappers. Same with IP-echo service.

Reply:
Hmm.. So
1. you _connect_ to someone who you know and who knows you
1. you bootstrap or ip-echo-req someone who you know but who doesn't know you

However `1` could lead to some rejections in connections esp if the other peer hasn't yet realised it needs you but you have. Eventually it will be connected because when it realises it will connect to you who has already realised it. This will need the upper layer to tell crust from whom it can _accept_ a direct connection request which is currently not the behaviour in Routing.

Let's actually leave this just now and discuss the intended behaviour.

Reply:
So to be clear: for this change-set go with what you were doing - leave the overall behaviour unchanged.